apps/lib/apps.c

```diff
@@ -2370,41 +2366,60 @@ int raw_write_stdout(const void *buf, int siz)
  * does impact behavior on some platforms, such as differentiating between
  * text and binary input/output on non-Unix platforms
  */
+static BIO *check_bio(BIO *bp, const char *what, int mode)
+{
+    if (bp == NULL) {
+        BIO_printf(bio_err, "%s: Could not open %s in mode 0x%x\n",
+                opt_getprog(), what, mode);
```

Comment:
I disagree, the flags are a bitmask.

Reply:
It depends. In `check_bio(ret, mode == 'r' ? "stdin" : "stdout", mode);` you are using the parameter for the fopen mode, which is a `char`.

Reply:
In the `check_bio(b, "stdout", format)` call in `bio_open_default()` 
the `flags` parameter is now abused in new way: to hold the format.

Reply:
I don't understand this comment.

Reply:
Having a look at
```
BIO *bio_open_default(const char *filename, char mode, int format)
{
    BIO *b;

    /* Using stdin or stdout? */
    if (filename == NULL || strcmp(filename, "-") == 0)
        return mode == 'r' ? dup_bio_in(format) : dup_bio_out(format);

    b = BIO_new_file(filename, modestr(mode, format));
    if (mode == 'r')
        return check_bio(b, "stdin", format);
    return check_bio(b, "stdout", format);
```
the last two lines call `check_bio(...,  ..., format)` where the last argument is not some type of flags as for the other calls, but the file contents format specifier, such at `FORMAT_PEM`, which is not a set of flags. 

To cure this new type of mix-up, I suggest replacing these two lines by 
```
check_bio(ret, mode == 'r' ? "stdin" : "stdout",
               mode == 'r' ? O_RDONLY : O_WRONLY);
```

Reply:
But that change is *wrong*  It's the ```FORMAT_xxx``` flags.

Reply:
Sorry, but you still seem to confuse/equate the `int flags` parameter of
```
BIO *BIO_new_fp(FILE *stream, int flags);

BIO_new_fp() creates a file BIO wrapping stream. Flags can be: BIO_CLOSE, BIO_NOCLOSE (the close flag) BIO_FP_TEXT
       (sets the underlying stream to text mode, default is binary: this only has any effect under Win32).
```
and the `int format` parameter of many file read/write functions, being the output of
```
int opt_format(const char *s, unsigned long flags, int *result)
```
which are two somewhat related but still different things.

For instance, `BIO_FP_TEXT ==  0x10` while `FORMAT_TEXT == 0x8000`.

Reply:
The issue, I guess, is that check_bio's diagnostic print is getting two different flags semantics.  So printing it as hex is good enough.  The alternative is having two/three check functions,  1 for fopen mode, 2 for open flags, 3 for BIO flags.  That's not worth doing in my view. Or to use the common phrase "for another PR"

Reply:
> The issue, I guess, is that check_bio's diagnostic print is getting two different flags semantics.

Right, `check_bio()` for the most part gets handed in BIO opening flags such as `BIO_NOCLOSE` (though they are still confusingly called 'mode' as mentioned in my other comment above) while via `bio_open_default()` at this point `check_bio()` gets handed in a format (i.e, file content type information such as `FORMAT_PKCS12`).

> So printing it as hex is good enough.

The diagnostic message `Could not open BIO(%s, 0x%x)` meanwhile lets it open how to interpret the hex value printed. This is not wrong but also not very helpful to the user because (s)he cannot know which way the value printed should be interpreted.

> The alternative is having two/three check functions, 1 for fopen mode, 2 for open flags, 3 for BIO flags.

The alternative I proposed would have been simpler and been to the point, namely to let the 
`check_bio()` parameter `flags` always be a BIO opening flag, and this could have been achieved simply by changing 
```
    if (mode == 'r')
        return check_bio(b, "stdin", format);
    return check_bio(b, "stdout", format);
```
to
```
check_bio(ret, mode == 'r' ? "stdin" : "stdout",
               mode == 'r' ? O_RDONLY : O_WRONLY);
```
Then the diagnostic message of `check_bio()` could be more informative and non-confusing: 
`Could not open BIO(filename=%s, flags=0x%x)`.

> That's not worth doing in my view. Or to use the common phrase "for another PR"

I agree that this can be sorted out later and is not worth the time to be discussed further in this PR.

Reply:
Opened https://github.com/openssl/openssl/issues/12056 to capture this.

Reply:
Fine